### ISSUE TITLE
Discovery.Manager not closing `syncCh` after sending stopped

### ISSUE THREAD

**Issue by littlepangdi** [prometheus/prometheus#10217]
**What did you do?**

`Loki` project quoting packages of `prometheus/discovery` for service discovery,and finds that after `discovery.Manager` exit, there is one unnormal go routine left due to an unclosed channel in Manager struct.

**What did you expect to see?**
After discussing with contributors in loki project(see more in[loki-pr-5238](https://github.com/grafana/loki/pull/5238)), 

We think it would be better the channel is closed within lifecycle of `discovery.Manager`, preferably when write process is over,for two reasons:

1.  in `documentation/examples/custom-sd/adapter/adapter.go` We give an example of reading from `Manager.Synch()`, and it will only exit after the channel is closed, see below

``` golang
func (a *Adapter) runCustomSD(ctx context.Context) {
	updates := a.manager.SyncCh()
	for {
		select {
		case <-ctx.Done():
		case allTargetGroups, ok := <-updates:
			// Handle the case that a target provider exits and closes the channel
			// before the context is done.
			if !ok {
				return
			}
			a.refreshTargetGroups(allTargetGroups)
		}
	}
}
```
2. If not, readers will have no choice but to shutdown goroutine via `ctx.Done()`, consider this situation casing messages losing:

* `ctx` is cancelled before readers read in all messages

 

**To Reproduce:**

1. call `discovery.NewManager(ctx)` and name it `m` in project
2. call `m.Synch()` to get a read copy of channel, and read from it 
3. call `ctx.CancelFunc()` 
4. find that `m.sender()`is stopped while not close the channel.

**Comment by roidelapluie** [comment]
I agree with you. however, we have to be very careful with this as we ave to check we never send anything to that closed channel.

**Comment by littlepangdi** [comment]
> 

Of course, I checked related code and found that there is a great design for this channel——send ops only happens within  goroutine of `sender()`，any other place references the channel is via read-mode, like :
https://github.com/prometheus/prometheus/blob/4cc25c0cb0b96042a7d36a0dd53dc6970ad607fd/cmd/prometheus/main.go#L1542

 so I suggested we can close it after `sender()` is closed natually. see in  #10218

**Comment by roidelapluie** [comment]
this PR is to fix #10217 , a possible memory leak when quoating manager.Sync() outside manager's lifecycle
also see more discuss in grafana/loki#5238

There is a great design for this channel——send ops only happens within goroutine of sender()，any other place references the channel is via read-mode, see below :
prometheus/cmd/prometheus/main.go

Line 1542 in 4cc25c0

 SyncCh() <-chan map[string][]*targetgroup.Group 
so I suggested we can close it after sender() is stopped naturally( in defer process,so there is no worry about sending to a closed channel).

**Comment by roidelapluie** [comment]
I have been investigating this deeply and this change is not good to me:

Prometheus is using this channel into the NotifyManager and the ScrapeManager. Both are reading from this channel. If this channel is closed, they would receive in loops zero-values from that channel.

The Prometheus server is using https://github.com/oklog/run to manage those goroutines. This packages does not guarantee any shutdown order, which means that the changes that Notify/Scrape manager might read un-necessarily zeroed values is not nil.

Therefore, I'd prefer to stick to the current behavior.

**Comment by beorn7** [comment]
Thanks for the PR. @roidelapluie would be best to look at it (but I nominated @bboreham as a backup ;).
